File: Perl/Math.pl
#! /usr/local/bin/perl

use Math::Trig; # pi に必要
use List::Util "max";
use List::Util "min";
use POSIX;      # log10 に必要
use Math::Complex;

print <<"Math";

これから数学演算を試します

進数変換
dec2bin 138: @{[   sprintf("%b",138)   ]}
dec2oct 138: @{[   sprintf("%o",138)   ]}
dec2hex 138: @{[   sprintf("%X",138)   ]}

dec2bin 138: @{[   sprintf("%#b",138)   ]}
dec2oct 138: @{[   sprintf("%#o",138)   ]}
dec2hex 138: @{[   sprintf("%#X",138)   ]}

bin2dec 1101111101: @{[   oct("0b1101111101")   ]}
oct2dec       1575: @{[   oct("1575")           ]}
hex2dec        37d: @{[   hex("37d")            ]}

定数
π = @{[pi]}     (use Math::Trig が必要)
e = @{[exp(1)]}

最大/最小
max(3,4,6) = @{[   max(3,4,6)   ]}
min(3,4,6) = @{[   min(3,4,6)   ]}

符号
abs(+18) = @{[   abs(+18)   ]}
abs(-18) = @{[   abs(-18)   ]}

数学関数
sqrt(3)     = @{[   sqrt(3)       ]}
exp(1.145)  = @{[   exp(1.145)    ]}
log(23.14)  = @{[   log(23.14)    ]}
log10(1385) = @{[   log10(1385)   ]} (use POSIX が必要)
sin(π/3)    = @{[   sin(pi/3)     ]}
cos(π/3)    = @{[   cos(pi/3)     ]}
tan(π/3)    = @{[   tan(pi/3)     ]}
asin(1)     = @{[   asin(1)       ]}
acos(0.5)   = @{[   acos(0.5)     ]}
atan(-1)    = @{[   atan(-1)      ]}
atan2(12,5) = @{[   atan2(12,5)   ]} atan2(y,x)
sinh(1.862) = @{[   sinh(1.862)   ]}
cosh(1.811) = @{[   cosh(1.811)   ]}
tanh(0.325) = @{[   tanh(0.325)   ]}

複素数
cplx(3,4)        = @{[   cplx(3,4)          ]}
cplxe(2,π/3)     = @{[   cplxe(2,pi/3)      ]}
(3+4*i)->Re()    = @{[   (3+4*i)->Re()      ]}
(3+4*i)->Im()    = @{[   (3+4*i)->Im()      ]}
(3+4*i)->abs()   = @{[   (3+4*i)->abs()     ]}
(3+4*i)->arg()   = @{[   (3+4*i)->arg()     ]}
(3+4*i)->rho()   = @{[   (3+4*i)->rho()     ]}
(3+4*i)->theta() = @{[   (3+4*i)->theta()   ]}
Math
=pod
   複素数型は成分毎に書き換え可能(mutable)
   e.g. $z->Im(5) : 複素数 $z の虚部を5に変更する
   数学関数においてこれらの複素数を利用できる
=cut
print <<"Math";
乱数
rand()          = @{[   rand()      ]} (0≦x<1)
rand(100)       = @{[   rand(100)   ]} (0≦x<100)

端数処理
floor(-3.14) = @{[   floor(-3.14)   ]} (小さい方の整数へ)
 ceil(-3.14) = @{[    ceil(-3.14)   ]} (大きい方の整数へ)
  int(-3.14) = @{[     int(-3.14)   ]} (キャストは0に近い方へ)
Math

print "\r\n\r\n";

File: Ruby/Math.rb
#! /usr/local/bin/ruby

print """

これから数学演算を試します

進数変換
433045 = #{"433045".to_i(6)} (6進数 → 10進数)
35669  = #{35669.to_s(14)}  (10進数 → 14進数)
""";
=begin
    2~36進数に変換可能
    to_i : 整数に変換
    to_f : 浮動小数に変換
    to_c : 複素数に変換
    to_r : 有理数に変換
    to_s : 文字列に変換
=end
print """
dec2bin 138: #{   sprintf("%b",138)   }
dec2oct 138: #{   sprintf("%o",138)   }
dec2hex 138: #{   sprintf("%X",138)   }

dec2bin 138: #{   sprintf("%#b",138)   }
dec2oct 138: #{   sprintf("%#o",138)   }
dec2hex 138: #{   sprintf("%#X",138)   }

bin2dec 1101111101: #{   "0b1101111101".oct   }
oct2dec       1575: #{   "1575".oct           }
hex2dec        37d: #{   "37d".hex            }

定数
π = #{Math::PI}
e = #{Math::E}

最大/最小
[3,4,6].min    = #{   [3,4,6].min      }
[3,4,6].max    = #{   [3,4,6].max      }
[3,4,6].minmax = #{   [3,4,6].minmax   }

符号
+18.abs = #{   +18.abs   }
-18.abs = #{   -18.abs   }

数学関数
Math.sqrt(3)     = #{   Math.sqrt(3)           }
Math.exp(1.145)  = #{   Math.exp(1.145)        }
Math.log(23.14)  = #{   Math.log(23.14)        }
Math.log10(1385) = #{   Math.log10(1385)       }
Math.sin(π/3)    = #{   Math.sin(Math::PI/3)   }
Math.cos(π/3)    = #{   Math.sin(Math::PI/3)   }
Math.tan(π/3)    = #{   Math.tan(Math::PI/3)   }
Math.asin(1)     = #{   Math.asin(1)           }
Math.acos(0.5)   = #{   Math.acos(0.5)         }
Math.atan(-1)    = #{   Math.atan(-1)          }
Math.atan2(12,5) = #{   Math.atan2(12,5)       } (atan2(y,x))
Math.sinh(1.862) = #{   Math.sinh(1.862)       }
Math.cosh(1.811) = #{   Math.cosh(1.811)       }
Math.tanh(0.325) = #{   Math.tanh(0.325)       }

複素数
Complex(3,4) = #{   Complex(3,4)   }
(3+4i).real  = #{   (3+4i).real    }
(3+4i).imag  = #{   (3+4i).imag    }
(3+4i).conj  = #{   (3+4i).conj    }
(3+4i).abs   = #{   (3+4i).abs     }
(3+4i).arg   = #{   (3+4i).arg     }
(3+4i).polar = #{   (3+4i).polar   }
(3+4i).rect  = #{   (3+4i).rect    }
""";
=begin
   複素数型は書き換えできない
   Mathに含まれる数学関数は複素数に対応していない
=end
print """

有理数
Rational(\"32/29\") = #{   Rational("32/29")    }
Rational(156,-84) = #{   Rational(156,-84)    }
Rational(π)       = #{   Rational(Math::PI)   }

Rational(\"32/29\").numerator   = #{   Rational("32/29").numerator     }
Rational(\"32/29\").denominator = #{   Rational("32/29").denominator   }

乱数
Random.rand            = #{   Random.rand              } (0≦x<1)
Random.rand(100)       = #{   Random.rand(100)         } (0≦x<100)
Random.rand(2.7..5.4)  = #{   Random.rand(2.7..5.4)    } (2.7≦x<5.4)
Random.rand(2.7...5.4) = #{   Random.rand(2.7...5.4)   } (2.7≦x≦5.4)

端数処理
-3.14.floor    = #{   -3.14.floor   } (小さい方の整数へ)
-3.14.ceil     = #{   -3.14.ceil    } (大きい方の整数へ)
-3.14.truncate = #{   -3.14.ceil    } (0に近い方の整数へ)
-3.14.round    = #{   -3.14.round   }
-3.14.to_i     = #{   -3.14.to_i    } (キャストはtruncateする)
""";

print "\r\n\r\n";